README
# NifMAP

[![Build Status](https://travis-ci.org/roey-angel/NifMAP.svg?branch=master)](https://travis-ci.org/roey-angel/NifMAP)

NifH MiSeq Illumina amplicon Analysis Pipeline 


changelog.txt
# Change Log

## [1.2] - 2021-04-28

### Added
1. NifMAP now works with usearch to merge reads, perform quality control, cluster OTUs and generate an OTU table.
2. nifH_bch_hmmEvaluation.R has been cosmetically updated but is still very slow for large datasets (no surpise).


NifMAP1.0.sh
#!/bin/bash -
#title          :NifMAP.sh
#description    :NifMAP - a bioinformatics pipline for analysing nifH amplicon data
#author         :Roey Angel
#date           :20180122
#version        :1.0
#usage          :./nifmap.sh $inputReads.fa
#notes          :This pipeline is designed to work in concert with the pipeline described by Herbold et al. 2015. See Angel et al. 2017 for details.
#dependencies   : 1. HMMER - http://hmmer.org/
#                 2. FrameBot - https://github.com/rdpstaff/Framebot
#                 3. seqmagick - https://fhcrc.github.io/seqmagick/
#                 4. CART - https://wwwzehr.pmc.ucsc.edu/CART_model_public/
#                 5. R - https://www.r-project.org/
#bash_version   :4.3.48(1)-release
#============================================================================

# 1. Merge MiSeq aplicon reads

# 2. Quality filter merged reads

# 3. Filter sequences using nifH HMM
# Define variables
inputReads=$1 # input fasta file containing merged MiSeq reads
HOMEFOLDER=`pwd`
FRAMEBOTPATH="/apps/rdptools/20160204"
WORKFOLDER=/tmp/$USER/NifMAP_work
RESULTSFOLDER=/tmp/$USER/NifMAP_results
RESOURCEFOLDER=$HOMEFOLDER/Resources

if [ ! -d /tmp/$USER ]
then
 mkdir /tmp/$USER
fi
if [ -d ${WORKFOLDER} ]
then
 rm -r ${WORKFOLDER}
fi
if [ -d ${RESULTSFOLDER} ]
then
 rm -r ${RESULTSFOLDER}
fi
mkdir ${WORKFOLDER}
mkdir ${RESULTSFOLDER}
cp ${HOMEFOLDER}/${inputReads} ${WORKFOLDER}

# Screen merged reads using HMMer
hmmsearch --domtblout ${WORKFOLDER}/hmmOut.out -o ${WORKFOLDER}/junk ${RESOURCEFOLDER}/hmm_nuc_1160_nifH.hmm ${WORKFOLDER}/${inputReads}
awk '{print $1}' ${WORKFOLDER}/hmmOut.out | grep -v "#" > ${WORKFOLDER}/acceptable_hits
grep ">" ${WORKFOLDER}/${inputReads} | grep -v -F -f ${WORKFOLDER}/acceptable_hits >${WORKFOLDER}/shitHits
totalUnique=`grep ">" ${WORKFOLDER}/${inputReads} | wc -l`
totalAccepted=`cat ${WORKFOLDER}/acceptable_hits | wc -l`
totalRemoved=`cat ${WORKFOLDER}/shitHits | wc -l`
echo "hmmscreen of nifH removed ${totalRemoved} sequences out of ${totalUnique} sequences. ${totalAccepted} unique sequences retained"> ${RESULTSFOLDER}/NifMAP_log.txt
mv ${WORKFOLDER}/${inputReads} ${WORKFOLDER}/${inputReads%.fa}_prehmm.fa
awk 'BEGIN{FS="\n";RS=">"};NR>1{print(">"$1);for(i=2;i<=NF;i++){printf($i)};print("")}' ${WORKFOLDER}/${inputReads%.fa}_prehmm.fa | grep -A 1 -F -f ${WORKFOLDER}/acceptable_hits | grep -v "^\-\-$" >${RESULTSFOLDER}/${inputReads}

# 4. Generate OTUs

# 5.Translate and correct OTU representatives using Framebot
eVal_chL=1e-50
score_chL=150
eVal_bchX=1e-50
score_bchX=150

cd ${WORKFOLDER}
java -jar ${FRAMEBOTPATH}/FrameBot.jar framebot -N -l 30 -i 0.4 -o ${inputReads%.fa} ${FRAMEBOTPATH}/Framebot/refset/nifh_prot_ref.fasta ${RESULTSFOLDER}/${inputReads}

# 6. Filter out homologous genes (bchL; chlL; bchX; parA) using HMM:
# Corrected AA sequences are in ${inputReads%.fa}_corr_prot.fasta
# Screen with hmm to identify all hits
hmmscan --domtblout ${WORKFOLDER}/hmmOut.out -o ${WORKFOLDER}/junk_NifH_ChlL_bchX ${RESOURCEFOLDER}/nifH_chlL_bchX.hmm ${WORKFOLDER}/${inputReads%.fa}_corr_prot.fasta
Rscript --vanilla < ${HOMEFOLDER}/nifH_bch_hmmEvaluation.R ${WORKFOLDER}/hmmOut.out
mv nifH_bch_hmmEvaluation.pdf ${RESULTSFOLDER}
cat hmmOut.out | awk 'NR>3{if($8>bitarray[$4]){bitarray[$4]=$8;outArray[$4]=$1"\t"$4}}END{for(entry in outArray){print outArray[entry]}}' > assignments.txt
cp hmmOut.out ${RESULTSFOLDER}/nifH_bch_hmmEvaluation.hmm.out
grep "nifH" assignments.txt | awk '{print $2"_"}' | sort > ${WORKFOLDER}/acceptable_hits
grep ">" ${RESULTSFOLDER}/${inputReads} | awk '{print $1"_"}' | grep -v -F -f ${WORKFOLDER}/acceptable_hits | sed 's/>//'>${WORKFOLDER}/shitHits
totalOTUs=`grep ">" ${RESULTSFOLDER}/${inputReads} | wc -l`
totalAccepted=`cat ${WORKFOLDER}/acceptable_hits | wc -l`
totalRemoved=`cat ${WORKFOLDER}/shitHits | wc -l`
echo "FRAMEBOT and hmmscreen of ${GENE} removed ${totalRemoved} sequences out of ${totalOTUs} OTU sequences. ${totalAccepted} otu reps retained">>${RESULTSFOLDER}/NifMAP_log.txt

mv ${RESULTSFOLDER}/${inputReads} ${WORKFOLDER}/${inputReads}
cat ${WORKFOLDER}/${inputReads} | sed 's/ //g' | awk 'BEGIN{RS=">";FS="\n"};NR>1{printf $1"_\t";for(i=2;i<=NF;i++){printf($i)}print("")}' | grep -F -f ${WORKFOLDER}/acceptable_hits | awk '{gsub("_$","",$1);print(">"$1"\n"$2)}' > ${RESULTSFOLDER}/${inputReads}
cat ${WORKFOLDER}/${inputReads%.fa}_corr_prot.fasta | sed 's/ //g' |  awk 'BEGIN{RS=">";FS="\n"};NR>1{printf $1"_\t";for(i=2;i<=NF;i++){printf($i)}print("")}' | grep -F -f ${WORKFOLDER}/acceptable_hits | awk '{gsub("_$","",$1);print(">"$1"\n"$2)}' > ${RESULTSFOLDER}/${inputReads%.fa}_corr_prot.fasta
cat ${WORKFOLDER}/${inputReads%.fa}_corr_prot.fasta | sed 's/ //g' |  awk 'BEGIN{RS=">";FS="\n"};NR>1{printf $1"_\t";for(i=2;i<=NF;i++){printf($i)}print("")}' | grep -F -f ${WORKFOLDER}/shitHits | awk '{gsub("_$","",$1);print(">"$1"\n"$2)}' > ${RESULTSFOLDER}/${inputReads%.fa}_rej_prot.fasta

# 7. Classify sequences using BLASTP, CART and RAxML-EPA
#RAxML-EPA
INPUTREFERENCEALIGNMENT=cluster.rep.nifH-chL-bchX.0.9_noDot.fasta.gz
INPUTREFERENCETREE=RAxML_bipartitions.MultipleOriginal.tree.gz
cp ${RESULTSFOLDER}/${inputReads%.fa}_corr_prot.fasta ./
cp ${RESOURCEFOLDER}/${INPUTREFERENCEALIGNMENT} ./
cp ${RESOURCEFOLDER}/${INPUTREFERENCETREE} ./

gunzip ${INPUTREFERENCEALIGNMENT}
gunzip ${INPUTREFERENCETREE}

mafft --add ${inputReads%.fa}_corr_prot.fasta --thread 16 ${INPUTREFERENCEALIGNMENT%.gz} > completeAlignment.fa

cat completeAlignment.fa | awk '{print $1}' | sed 's/|$//'> RAxml.compatible.fa
raxmlHPC-PTHREADS-SSE3 -f v -T 16 -s RAxml.compatible.fa -m PROTCATJTT -t RAxML_bipartitions.MultipleOriginal.tree -n EPAplaced -p 123
mkdir EPA
cp *EPAplaced* EPA
cp completeAlignment.fa EPA
mv EPA ${RESULTSFOLDER}

# CART
cat ${RESOURCEFOLDER}/CART_nifH/AztVine8_AA.fasta ${RESULTSFOLDER}/${inputReads%.fa}_corr_prot.fasta > ${WORKFOLDER}/${inputReads%.fa}_corr_prot_4classification.fa
hmmalign --amino ${RESOURCEFOLDER}/Zehr_2014_1812genomes_nifH_AA_noGaps.hmm ${WORKFOLDER}/${inputReads%.fa}_corr_prot_4classification.fa > ${WORKFOLDER}/${inputReads%.fa}_corr_prot_hmm.sth
seqmagick convert  ${WORKFOLDER}/${inputReads%.fa}_corr_prot_hmm.sth  ${WORKFOLDER}/${inputReads%.fa}_corr_prot_hmmAln.fasta
python ${RESOURCEFOLDER}/CART_nifH/NifH_Clusters.py ${WORKFOLDER}/${inputReads%.fa}_corr_prot_hmmAln 1
grep '>' ${WORKFOLDER}/${inputReads%.fa}_corr_prot_hmmAln_Clusters.fasta |  sed 's/>//' | awk '{print $1, $5, $8}' > ${RESULTSFOLDER}/${inputReads%.fa}_corr_prot.zehr.classification

# BLASTP
wget http://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz ./
tar -zxvf taxdb.tar.gz
blastp -query ${RESULTSFOLDER}/${inputReads%.fa}_corr_prot.fasta -db /localmirror/monthly/blast/ncbi/refseq_protein -num_threads 16 -max_target_seqs 1 -out ${RESULTSFOLDER}/${inputReads%.fa}_refseq-prot_nifH.out -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore staxids sscinames sblastnames sskingdoms stitle"

# 8. Return results and clean up
#return results
cp ${RESULTSFOLDER} $HOMEFOLDER/

#cleanup
rm -r ${WORKFOLDER}
rm -r ${RESULTSFOLDER}


nifH_bch_hmmEvaluation.R
# Plot HMM output
# This is probably the most inefficient code I ever wrote,
#   but it works so I'm unmotivated to change it :(
# V0.1

install.packages("ggplot2", lib = "./", repos = "http://cran.wu.ac.at")
library(ggplot2,lib.loc = "./")
install.packages("vegan", lib = "./", repos = "http://cran.wu.ac.at")
library(vegan,lib.loc = "./")

PlotHmmOutput <- function(HMM.OUT, type = "eval", gene = "bchX") {
  require(ggplot2)
  HMM.OUT$V24 <- log(HMM.OUT$V7)
  if (type == "eval") {
    ggplot(HMM.OUT, aes(V24)) +
      geom_freqpoly(binwidth = 2) +
      geom_vline (xintercept = log(1e-50), colour = "red") +
      labs(x = "ln E-value", y = "Count", title = paste(gene, "HMM model E-value distribution"))
  } else {
    ggplot(HMM.OUT, aes(V8)) +
      geom_freqpoly(binwidth = 2) +
      geom_vline (xintercept = 150, colour = "red") +
      labs(x = "Score", y = "Count", title = paste(gene, "HMM model Score distribution"))
  }
}

pdf("nifH_bch_hmmEvaluation.pdf")
HMM.OUT <- read.table(commandArgs()[3])
PlotHmmOutput(HMM.OUT[grep("nifH", HMM.OUT[,1]),], 
              type = "eval", gene = "nifH")
PlotHmmOutput(HMM.OUT[grep("nifH", HMM.OUT[,1]),], 
              type = "score",
              gene = "nifH")
PlotHmmOutput(HMM.OUT[HMM.OUT[, 1] == "bchX", ], 
              type = "eval", gene = "bchX")
PlotHmmOutput(HMM.OUT[HMM.OUT[, 1] == "bchX", ], 
              type = "score",
              gene = "bchX")
PlotHmmOutput(HMM.OUT[HMM.OUT[, 1] == "chlL-bchL", ],
              type = "eval", gene = "chlL-bchL")
PlotHmmOutput(HMM.OUT[HMM.OUT[, 1] == "chlL-bchL", ], 
              type = "score", 
              gene = "chlL-bchL")

OTUnums <- sort(as.numeric(gsub("[^0-9]+", "", unique(HMM.OUT[, 4]))))
scoreTable <- matrix(0, nrow = length(OTUnums),ncol = 3)
rownames(scoreTable) <- OTUnums
colnames(scoreTable) <- c("nifH", "chlL", "bchX")
evalTable <- scoreTable
nifHSet <- grep("nifH", HMM.OUT[, 1])
chlLSet <- grep("chlL", HMM.OUT[, 1])
bchXSet <- grep("bchX", HMM.OUT[, 1])
for(i in seq(nrow(scoreTable))){
	thisOTUset <- which(as.numeric(gsub("[^0-9]+", "", HMM.OUT[, 4])) == rownames(scoreTable)[i])
	scoreTable[i, 1] <- HMM.OUT[intersect(thisOTUset, nifHSet)[1], 8]
	scoreTable[i, 2] <- HMM.OUT[intersect(thisOTUset, chlLSet)[1], 8]
	scoreTable[i, 3] <- HMM.OUT[intersect(thisOTUset, bchXSet)[1], 8]
	evalTable[i, 1] <- HMM.OUT[intersect(thisOTUset, nifHSet)[1], 7]
	evalTable[i, 2] <- HMM.OUT[intersect(thisOTUset, chlLSet)[1], 7]
	evalTable[i, 3] <- HMM.OUT[intersect(thisOTUset, bchXSet)[1], 7]

}
assignments <- rownames(which(apply(scoreTable, 1, rank) == 3, arr.ind = TRUE))

par(mfrow = c(1, 1))
plot(scoreTable[, 2], 
     scoreTable[, 1],
     pch = 16,
     col = as.numeric(as.factor(assignments)),
     xlab = colnames(scoreTable)[2],
     ylab = "nifH",
     main = "separation of ChlL")
abline(a = 0, 
       b = 1)
legend("topleft",
       levels(as.factor(assignments)),
       col = c(1:length(levels(as.factor(assignments)))), 
       pch = 16)
plot(scoreTable[, 3], 
     scoreTable[, 1],
     pch = 16,
     col = as.numeric(as.factor(assignments)),
     xlab = colnames(scoreTable)[3],
     ylab = "nifH", 
     main = "separation of BchX")
abline(a = 0, 
       b = 1)
legend("topleft",
       levels(as.factor(assignments)),
       col = c(1:length(levels(as.factor(assignments)))),
       pch = 16)

pcoA <- rda(scoreTable)
plot(pcoA$CA$u[, 1] * pcoA$CA$eig[1],
     pcoA$CA$u[, 2] * pcoA$CA$eig[2],
     col = as.numeric(as.factor(assignments)),
     pch = 16,
     xlab = paste("PcoA 1 (", round(pcoA$CA$eig[1] / sum(pcoA$CA$eig), 4) * 100, "%)"),
     ylab = paste("PcoA 2 (", round(pcoA$CA$eig[2] / sum(pcoA$CA$eig), 4) * 100, "%)"))
legend("topleft",
       levels(as.factor(assignments)),
       col = c(1:length(levels(as.factor(assignments)))),
       pch = 16)
dev.off()


NifMAP1.2.sh
#!/bin/bash -
#title          :NifMAP.sh
#description    :NifMAP - a bioinformatics pipline for analysing nifH amplicon data
#author         :Roey Angel
#date           :20210309
#version        :1.2
#usage          :./nifmap.sh <inputDir>
#notes          : Unlike V1.0, this pipeline is now independent and performs quality-filtering and OTU generation.
#               See Angel, R., Nepel, M., Panhölzl, C., Schmidt, H., Herbold, C. W., Eichorst, S. A., et al. (2018). Evaluation of primers targeting the diazotroph functional gene and development of NifMAP – a bioinformatics pipeline for analyzing nifH amplicon data. Front Microbiol 9, 703. doi:10.3389/fmicb.2018.00703.
#dependencies   : 1. HMMER - http://hmmer.org/
#                 2. FrameBot - https://github.com/rdpstaff/Framebot
#                 3. seqmagick - https://fhcrc.github.io/seqmagick/
#                 4. CART - https://wwwzehr.pmc.ucsc.edu/CART_model_public/
#                 5. R - https://www.r-project.org/
#bash_version   :4.3.48(1)-release
#============================================================================

eval "$(conda shell.bash hook)" # this is needed to be able to use conda activate

# Define variables
inputDir="./Data" #$1 # input fasta file containing merged MiSeq reads
HOMEFOLDER=`pwd` # base library
FRAMEBOTPATH="/usr/local/lib/Bioinformatics/rdptools/2.0.2/"
WORKFOLDER=${HOMEFOLDER}/nifH_work
RESULTSFOLDER=${HOMEFOLDER}/nifH_results
RESOURCEFOLDER=${HOMEFOLDER}/Resources
INPUTREFERENCEALIGNMENT=cluster.rep.nifH-chL-bchX.0.9_noDot.fasta
INPUTREFERENCETREE=RAxML_bipartitions.MultipleOriginal.tree
id=`echo 0.94 | bc` # OTU clustering, 1.0 = zOTUS, 0.97 = 97%
cores=4 # number of cores to use
MINLEN=200 # merged reads with a smaller size will be removed

if [ -d ${WORKFOLDER} ]
then
 rm -r ${WORKFOLDER}
fi
if [ -d ${RESULTSFOLDER} ]
then
 rm -r ${RESULTSFOLDER}
fi
mkdir ${WORKFOLDER}
mkdir ${RESULTSFOLDER}

echo "NifMAP1.2"> ${RESULTSFOLDER}/NifMAP.log

# 1. Merge MiSeq aplicon reads
echo "usearch -fastq_mergepairs">> ${RESULTSFOLDER}/NifMAP.log
cd $inputDir
#for gzfile in *.fastq.gz; do; gunzip $gzfile; done
usearch -fastq_mergepairs *_R1.fastq -fastqout merged_reads.fq  -relabel @  -threads $cores --log ${RESULTSFOLDER}/tmp.log
cd ../
cat ${RESULTSFOLDER}/tmp.log >>  ${RESULTSFOLDER}/NifMAP.log
echo "--">> ${RESULTSFOLDER}/NifMAP.log

# 2. Quality filter merged reads
echo "usearch -fastq_filter">> ${RESULTSFOLDER}/NifMAP.log
#for inputReads in ${inputDir}/*.fastq; do
#    usearch -fastq_filter ${inputReads} -fastq_maxee 1.0 -fastq_minlen 200 -relabel @ -fastaout ${WORKFOLDER}/${inputReads%.*}.fasta --log ${RESULTSFOLDER}/tmp.log;
#    cat ${RESULTSFOLDER}/tmp.log >>  ${RESULTSFOLDER}/NifMAP.log
#done
usearch -fastq_filter ${inputDir}/merged_reads.fq -fastq_maxee 1.0 -fastq_minlen $MINLEN -fastaout ${WORKFOLDER}/merged_reads.fa -threads $cores --log ${RESULTSFOLDER}/tmp.log
cat ${RESULTSFOLDER}/tmp.log >>  ${RESULTSFOLDER}/NifMAP.log
echo "--" >> ${RESULTSFOLDER}/NifMAP.log

# 3. Filter sequences using nifH HMM
# Screen merged reads using HMMer
echo "Screen merged reads using HMMer">> ${RESULTSFOLDER}/NifMAP.log
hmmsearch --cpu $cores --domtblout ${WORKFOLDER}/hmmOut1.out -o ${WORKFOLDER}/junk ${RESOURCEFOLDER}/hmm_nuc_1160_nifH.hmm ${WORKFOLDER}/merged_reads.fa
awk '{print $1}' ${WORKFOLDER}/hmmOut1.out | grep -v "#" > ${WORKFOLDER}/acceptable_hits
grep ">" ${WORKFOLDER}/merged_reads.fa | grep -v -F -f ${WORKFOLDER}/acceptable_hits >${WORKFOLDER}/shitHits
totalUnique=`grep ">" ${WORKFOLDER}/merged_reads.fa | wc -l`
totalAccepted=`cat ${WORKFOLDER}/acceptable_hits | wc -l`
totalRemoved=`cat ${WORKFOLDER}/shitHits | wc -l`
echo "hmmscreen of nifH removed ${totalRemoved} sequences out of ${totalUnique} sequences. ${totalAccepted} unique sequences retained" >> ${RESULTSFOLDER}/NifMAP.log
mv ${WORKFOLDER}/merged_reads.fa ${WORKFOLDER}/merged_reads_prehmm.fa
awk 'BEGIN{FS="\n";RS=">"};NR>1{print(">"$1);for(i=2;i<=NF;i++){printf($i)};print("")}' ${WORKFOLDER}//merged_reads_prehmm.fa | grep -A 1 -F -f ${WORKFOLDER}/acceptable_hits | grep -v "^\-\-$" >${RESULTSFOLDER}/merged_reads.fa
echo "--">> ${RESULTSFOLDER}/NifMAP.log

# 4. Generate OTUs
# Dereplicate sequences for OTU table
echo "usearch -fastx_uniques">> ${RESULTSFOLDER}/NifMAP.log
echo "Singletons will not be allowed to generate OTUs">> ${RESULTSFOLDER}/NifMAP.log
usearch -fastx_uniques ${RESULTSFOLDER}/merged_reads.fa -minuniquesize 2 -fastaout ${RESULTSFOLDER}/unique.fa -threads $cores --log ${RESULTSFOLDER}/tmp.log # -sizeout
cat ${RESULTSFOLDER}/tmp.log >>  ${RESULTSFOLDER}/NifMAP.log
echo "--">> ${RESULTSFOLDER}/NifMAP.log

# Make zOTUs or OTUs
if [[ (( $id == 1.0 )) ]]; then
    echo "usearch -unoise3 ">> ${RESULTSFOLDER}/NifMAP.log
    usearch -sortbylength ${RESULTSFOLDER}/unique.fa -fastaout ${WORKFOLDER}/unique_sorted.fa
    usearch -unoise3 ${WORKFOLDER}/unique_sorted.fa -zotus ${WORKFOLDER}/otus.fa -tabbedout ${RESULTSFOLDER}/unoise3.txt -threads $cores --log ${RESULTSFOLDER}/tmp.log
    cat ${RESULTSFOLDER}/tmp.log >>  ${RESULTSFOLDER}/NifMAP.log
    echo "--">> ${RESULTSFOLDER}/NifMAP.log
else
    echo "usearch -cluster_fast">> ${RESULTSFOLDER}/NifMAP.log
    usearch -sortbylength ${RESULTSFOLDER}/unique.fa -fastaout ${WORKFOLDER}/unique_sorted.fa
    usearch -cluster_fast ${WORKFOLDER}/unique_sorted.fa -id $id -centroids ${WORKFOLDER}/otus.fa -uc ${WORKFOLDER}/clusters.uc -threads $cores --log ${RESULTSFOLDER}/tmp.log
    cat ${RESULTSFOLDER}/tmp.log >>  ${RESULTSFOLDER}/NifMAP.log
    echo "--">> ${RESULTSFOLDER}/NifMAP.log
fi


# 5.Translate and correct OTU representatives using Framebot
#eVal_chL=1e-50
#score_chL=150
#eVal_bchX=1e-50
#score_bchX=150

java -jar ${FRAMEBOTPATH}/FrameBot.jar framebot -N -l 30 -i 0.4 -o ${WORKFOLDER}/nifH ${FRAMEBOTPATH}/Framebot/refset/nifh_prot_ref.fasta ${WORKFOLDER}/otus.fa
cp ${WORKFOLDER}/nifH_corr_nucl.fasta ${RESULTSFOLDER}


# 6. Make an OTU table based on corrected seqs
#echo "usearch -usearch_global ">> ${RESULTSFOLDER}/NifMAP.log
#usearch -usearch_global ${RESULTSFOLDER}/merged_reads.fa -db ${RESULTSFOLDER}/nifH_corr_nucl.fasta  -strand plus -id 0.97 -threads $cores -otutabout ${RESULTSFOLDER}/otu_table.txt --log ${RESULTSFOLDER}/tmp.log
#cat ${RESULTSFOLDER}/tmp.log >>  ${RESULTSFOLDER}/NifMAP.log
#echo "--">> ${RESULTSFOLDER}/NifMAP.log

# 7. Filter out homologous genes (bchL; chlL; bchX; parA) using HMM:
# Corrected AA sequences are in nifH_corr_prot.fasta
# Screen with hmm to identify all hits
hmmscan --cpu $cores  --domtblout ${WORKFOLDER}/hmmOut2.out -o ${WORKFOLDER}/full_output_NifH_ChlL_bchX ${RESOURCEFOLDER}/NifH_ChlL_bchX.hmm ${WORKFOLDER}/nifH_corr_prot.fasta
#Rscript --vanilla < ${HOMEFOLDER}/nifH_bch_hmmEvaluation.R ${WORKFOLDER}/hmmOut2.out
#mv nifH_bch_hmmEvaluation.pdf ${RESULTSFOLDER}
cat ${WORKFOLDER}/hmmOut2.out | awk 'NR>3{if($8>bitarray[$4]){bitarray[$4]=$8;outArray[$4]=$1"\t"$4}}END{for(entry in outArray){print outArray[entry]}}' > ${WORKFOLDER}/assignments.txt
cp ${WORKFOLDER}/hmmOut2.out ${RESULTSFOLDER}/nifH_bch_hmmEvaluation.hmm.out
grep "nifH" ${WORKFOLDER}/assignments.txt | awk '{print $2}' | sort > ${WORKFOLDER}/acceptable_hits
grep ">" ${RESULTSFOLDER}/nifH_corr_nucl.fasta | awk '{print $1}' | grep -v -F -f ${WORKFOLDER}/acceptable_hits | sed 's/>//'>${WORKFOLDER}/shitHits
totalOTUs=`grep ">" ${RESULTSFOLDER}/nifH_corr_nucl.fasta | wc -l`
totalAccepted=`cat ${WORKFOLDER}/acceptable_hits | wc -l`
totalRemoved=`cat ${WORKFOLDER}/shitHits | wc -l`
echo "FRAMEBOT and hmmscreen of nifH removed ${totalRemoved} sequences out of ${totalOTUs} ASVs. ${totalAccepted} ASV reps retained">>${RESULTSFOLDER}/NifMAP.log

#mv ${RESULTSFOLDER}/${inputReads} ${WORKFOLDER}/${inputReads}
cat ${RESULTSFOLDER}/nifH_corr_nucl.fasta | sed 's/ //g' | awk 'BEGIN{RS=">";FS="\n"};NR>1{printf $1"_\t";for(i=2;i<=NF;i++){printf($i)}print("")}' | grep -F -f ${WORKFOLDER}/acceptable_hits | awk '{gsub("_$","",$1);print(">"$1"\n"$2)}' > ${RESULTSFOLDER}/nifH_corr_nucl_only_nifH.fasta
cat ${WORKFOLDER}/nifH_corr_prot.fasta | sed 's/ //g' |  awk 'BEGIN{RS=">";FS="\n"};NR>1{printf $1"_\t";for(i=2;i<=NF;i++){printf($i)}print("")}' | grep -F -f ${WORKFOLDER}/acceptable_hits | awk '{gsub("_$","",$1);print(">"$1"\n"$2)}' > ${RESULTSFOLDER}/nifH_corr_prot.fasta
cat ${WORKFOLDER}/nifH_corr_prot.fasta | sed 's/ //g' |  awk 'BEGIN{RS=">";FS="\n"};NR>1{printf $1"_\t";for(i=2;i<=NF;i++){printf($i)}print("")}' | grep -F -f ${WORKFOLDER}/shitHits | awk '{gsub("_$","",$1);print(">"$1"\n"$2)}' > ${RESULTSFOLDER}/nifH_rej_prot.fasta

# Make an OTU table based on filter corrected seqs
echo "usearch -usearch_global ">> ${RESULTSFOLDER}/NifMAP.log
#usearch -usearch_global ${RESULTSFOLDER}/merged_reads.fa -db ${RESULTSFOLDER}/nifH_corr_nucl_only_nifH.fasta -id 1.0 -strand plus -threads $cores -otutabout ${RESULTSFOLDER}/otu_table.txt
usearch -otutab ${RESULTSFOLDER}/merged_reads.fa -otus ${RESULTSFOLDER}/nifH_corr_nucl_only_nifH.fasta -otutabout ${RESULTSFOLDER}/otu_table.txt -mapout ${WORKFOLDER}/map.txt -threads $cores

# 7. Place OTUs on a reference tree using RAxML-EPA
#RAxML-EPA
mkdir ${RESULTSFOLDER}/EPA
cp ${RESULTSFOLDER}/nifH_corr_prot.fasta ${RESULTSFOLDER}/EPA/
cp ${RESOURCEFOLDER}/${INPUTREFERENCEALIGNMENT} ${RESULTSFOLDER}/EPA/
cp ${RESOURCEFOLDER}/${INPUTREFERENCETREE} ${RESULTSFOLDER}/EPA/

cd ${RESULTSFOLDER}/EPA
mafft --add nifH_corr_prot.fasta --thread 16 ${INPUTREFERENCEALIGNMENT} > completeAlignment.fa
cat completeAlignment.fa | awk '{print $1}' | sed 's/|$//'> RAxML_compatible.fa
raxmlHPC-PTHREADS-SSE3 -f v -T $cores -s RAxML_compatible.fa -m PROTCATJTT -t ${INPUTREFERENCETREE} -n EPAplaced -p 123

cd ../../

# 8. Classify sequences using CART
cat ${RESOURCEFOLDER}/AztVine8_AA.fasta ${RESULTSFOLDER}/nifH_corr_prot.fasta > ${WORKFOLDER}/nifH_corr_prot_4classification.fa
hmmalign --amino ${RESOURCEFOLDER}/Zehr_2014_1812genomes_nifH_AA_noGaps.hmm ${WORKFOLDER}/nifH_corr_prot_4classification.fa > ${WORKFOLDER}/nifH_corr_prot_hmm.sth
conda activate base
seqmagick convert ${WORKFOLDER}/nifH_corr_prot_hmm.sth  ${WORKFOLDER}/nifH_corr_prot_hmmAln.fasta
conda deactivate
python2 ${RESOURCEFOLDER}/NifH_Clusters.py ${WORKFOLDER}/nifH_corr_prot_hmmAln 1
grep '>' ${WORKFOLDER}/nifH_corr_prot_hmmAln_Clusters.fasta |  sed 's/>//' | awk '{print $1, $5, $8}' > ${RESULTSFOLDER}/nifH_corr_prot.zehr.classification

# 9. Classify sequences using BLASTP
wget http://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz ./
tar -zxvf taxdb.tar.gz

blastp -query ${RESULTSFOLDER}/nifH_corr_prot.fasta -db /proj/Resources/BLAST/NCBI/refseq_protein -num_threads 80 -max_target_seqs 1 -out ${RESULTSFOLDER}/nifH_blastp.out -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore staxids sscinames sblastnames sskingdoms stitle"

# 8. Return results and clean up
#return results
#cp ${RESULTSFOLDER} $HOMEFOLDER/

#cleanup
rm -r ${RESULTSFOLDER}/tmp.log ${RESULTSFOLDER}/unique.fa ${RESULTSFOLDER}/unoise3.txt
rm -r ${WORKFOLDER}
rm taxdb*
